contributors/design-proposals/vertical-pod-autoscaler.md

```diff
@@ -0,0 +1,486 @@
+after the VPA object was added, in which case it’s not possible to predict
+whether the Pods will be scheduled using original resources or recommended
+resources.
+Both VPA Admission Controller and VPA Controller will remove the VPA initializer
+after setting Pod resources. Details of this mechanism will be designed when
+Initializers are available.
+
+### Admission Controller ###
+
+VPA Admission Controller intercepts Pod creation requests. If the Pod is matched
+by a VPA config with mode not set to “off”, the controller rewrites the request
+by applying recommended resources to the Pod spec. Otherwise it leaves the Pod
+spec unchanged.
+
+The controller gets the recommended resources by fetching
+/recommendedPodResources from the Recommender. If the call times out or fails,
+the controller falls back to the recommendation cached in the VPA object.
+If this is also not available and the Pod doesn’t have a VPA Initializer, it is
+rejected.
+
+### Life of the Pod ###
+There are two paths for a Pod using VPA to get scheduled.
+
+1. A synchronous (fast) path: if a matching VPA exists at the time of creating
+   the Pod and it is able to obtain a recommendation (either by calling the
+   Recommender or by using a cached value in the VPA resource), the Admission
+   Plugin sets resources synchronously. It also immediately deletes the VPA
+   initializer if one was present, to shorten the critical path for Pod
+   creation.
+
+2. An asynchronous (slower) path. A Pod is created with the Initializer.
+   The Admission Plugin doesn't see a matching VPA so it passes the Pod
+   unchanged. Once a matching VPA appears, the VPA controller sets the Pod
+   resources and removes the Initializer (one could also consider adding
+   optional deadline to the Initializer to let the Pod schedule if the VPA
+   doesn’t appear).
+
+![VPA Scheduling Flow](images/vpa-scheduling-flow.png "VPA scheduling flow")
+
+### Recommender ###
+Recommender is the main component of the VPA. It is responsible for
+computing recommended resources. On startup the recommender fetches
+historical resource utilization of all Pods (regardless of whether
+they use VPA) together with the history of Pod OOM events from the
+History Storage. It aggregates this data and keeps it in memory.
+
+During normal operation the recommender consumes real time updates of
+resource utilization and new events via the Master Metrics API from
+Heapster. Additionally it watches all Pods and all VPA objects in the
+cluster. For every Pod that is matched by some VPA selector the
+Recommender computes the recommended resources and sets the
+recommendation on the VPA object.
+
+It is important to realize that one VPA object has one recommendation.
+The user is expected to use one VPA to control Pods with similar
+resource usage patterns, typically a group of replicas or shards of
+a single workload.
+
+The Recommender exposes a synchronous RPC interface that takes a Pod
```

Comment:
My understanding is that Heapster is an exception only because Aggregated API Servers don't yet exist. Once they are available, Heapster will use this. And VPA recommendations API will be no different.